Volograms/vol_libs
# Big file (>2GB) support in vol_libs

**Reason for PR**

* Studio captures >2GB were not running in the Unity Plugin project. https://github.com/Volograms/volograms_unity_plugin
* Analysis showed that vol_libs was rejecting files of this size with an error message, and that underlying file size and seek functions used were not 64-bit offset aware on all platforms.

**Work Done in this PR**

* Functions `fseek()`, `ftell()` were replaced by fseeko(),ftello() on POSIX-compliant platforms, and Windows now uses `_fseeki64()`, `_ftelli64()`, which have equivalent functionality.
* A `#define` was introduced to differentiate platforms, which now in code refer to `vol_geom_fseeko` etc.
* Instead of using `fseek`/`ftell` for getting file size after opening the file, the functions `stat()/_stat64()` (Windows) are used, which does not require opening or seeking through large files.
* These functions are from `sys/stat.h` on both systems.
* Set `-D_FILE_OFFSET_BITS=64` in Makefile for POSIX builds, which enforces 64-bit `off_t` even on 32-bit systems (e.g. Nexus 6 or something). 
* Updated headers and definitions for Linux builds.
* Please note I also changed the default `vol_geom_size_t` type from `uint64_t` to `int64_t` to more closely match the above functions' offset types.
    * This may affect the Unity Plugin's library wrapper.
    * Because I have moved to a signed type, several places require a (type cast) to silence warnings of addition of bytes between the `vol_geom_size_t` and `size_t`.

**Testing Summary**

* Tested large captures with older C# player, to ensure they were playable and the files can work. (They worked).
* Tested on Windows with MinGW/GCC, and MSVC (2019) builds.
* Tested on Windows 10 with ld, sd, and hd captures from the "Ulysses" project. Before, and after this PR, using example programs in this library.
   * Smallest ld capture worked before the change, the others all worked after the changes in this PR.
   * Initial load times for very large files can be a few seconds.
   * Run-time memory footprint for ~8GB HD files is <60MB, and performance on desktop is great, using "streaming" mode. CPU usage on my desktop is very low. This is really good from a low-perf device point of view.
   * Run-time memory footprint for HD with 2k textures is ~90MB main memory. Also very low CPU footprint, but fairly high GPU usage. Hardware video decoding is not used, and very little GPU memory is retained.
![image](https://user-images.githubusercontent.com/1935602/170291788-b41422c7-9c4c-4d72-b27e-5950e93f5d7a.png)
![image](https://user-images.githubusercontent.com/1935602/170292763-7f570268-c66a-4050-b2f6-1ab98c66d765.png)
* Tested HD version on 64-bit Ubuntu (also used by GitHub Actions for running Tests).

**Risk Introduced by PR**

* *Regression risk of these changes is high*.
   * This library is a dependency of multiple projects and shipped products, so we'll need to be careful to test this very thoroughly, as we roll it out into other projects.
   * Testing on multiple platforms/compilers is essential.
   * Doing roll-out to other projects in stages (not all at once) is a good idea.
   * Update the Volu app last, to minimise risk to shipped product.
   * Any other very large files would be really handy to test with.
   
**ToDo (Before Merging)**

* Test on macOS ( **need a hand here!** )
   
**ToDo (After Merging)**

* Tag and Release 0.10.0
* Update Unity Plugin with latest vol_libs. (iOS and Android testing).
* Update Unreal Plugin with latest vol_libs.
* Update Volu app with latest vol_libs. (iOS and Android testing).
* Update vol_cl_tools with latest vol_libs.
* Lots of testing!
